### PR TITLE
Update to Pond V2

### PR DESCRIPTION
**Pull request by benjamin-actyx**
Update package to work with Pond 2.x

I split the registry into `fromEventTypes` and `generic`, in order to make it usable even when there is no `type` property on events.
(The naming and exports could use some improvement)

NB. the API is designed to play well with https://github.com/actyx-contrib/fish-type

Going to add a couple more tests

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -23,18 +23,35 @@ npm install @actyx-contrib/registry
...
+import { FishType } from '@actyx-contrib/fish-type-v2'
+import { fromTypes } from '@actyx-contrib/registry-v2'
+
+enum EventType {
+    create = 'create',
+    delete = 'delete',
+}
+
+type Event = { type: EventType, entityId: string }
+type State = { entityId: string, state: 'unknown' | 'created' }
```
Comment (aruediger) [inline README.md:38]:
```suggestion
type State = { entityId: string, type: 'unknown' | 'created' }
```

File: src/from-event-types.spec.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
+/* eslint-disable @typescript-eslint/no-use-before-define */
+// import { Observable } from 'rxjs'
```
Comment (aruediger) [inline src/from-event-types.spec.ts:2]:
```suggestion
```

File: src/from-event-types.spec.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+  it('should forget about fish', async () => {
+    const pond = await Pond.test()
+
+    const T = TestCtrl(pond)
+    T.setFish('foo')
+    T.setFish('bar')
+    T.deleteFish('foo')
+
+    return expect(T.obsAll('set', 'delete')).resolves.toEqual([{ fishName: 'bar', type: 'set' }])
```
Comment (aruediger) [inline src/from-event-types.spec.ts:93]: was kind of hard to spot the difference to the above test case. (mind the [].)

File: src/from-event-types.spec.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  extractFishName: (e: Event) => e.fishName,
+
+  events: TestTag,
+}
+
+const emitEvent = (pond: Pond, e: Event) =>
+  pond.emit<Event>(TestTag.withId(TestFish.extractFishName(e)), e)
+
+describe('registry from event types', () => {
+  const TestCtrl = (pond: Pond) => {
```
Comment (aruediger) [inline src/from-event-types.spec.ts:46]: could also create `pond` here

File: src/from-event-types.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+ * type Event = { type: EventType, entityId: string }
+ * const Events = Tag<Event>('example-tag')
+ *
+ * const RegistryFish1 = createRegistryFish(Events, (e: Event) => e.entityId, EventType.create)
+ * const RegistryFish2 = createRegistryFish(Events, (e: Event) => e.entityId, EventType.create, EventType.deleted)
+ * const RegistryFish3 = createRegistryFish(Events, (e: Event) => e.entityId, [EventType.create])
+ * const RegistryFish4 = createRegistryFish(Events, (e: Event) => e.entityId, [EventType.create], [EventType.deleted])
+ * ```
+ *
+ * @param events             Event stream to base the registry on
```
Comment (aruediger) [inline src/from-event-types.ts:44]:
```suggestion
 * @param events             Event stream to base the registry on.
```

File: src/generic-registry.ts [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+ * }
+ * const RegistryFish = createRegistryFish('my-registry', Events, (e: Event) => e.entityId, handler)
+ * ```
+ *
+ * @param uniqueDescriptor   Give your registry a name. Used for caching!
+ * @param events             Event stream to base the registry on.
+ * @param extractFishName    How to get the related Fish’s name from an event.
+ * @param handler            A function that decides whether a given event leads to addition or removal of its related entity.
+ */
+export const createRegistryFish = <E>(
```
Comment (aruediger) [inline src/generic-registry.ts:64]: should this be called `createRegistryFishGeneric`? i'd just drop the "Generic" suffix from `observeAllGeneric`.

File: src/types.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  descriptor: string
+
+  /** Create a Fish instance from a FishName. */
+  makeFish: MakeFish<Id, S>
+
+  /** From an event, extract the name of the Fish it belongs to. */
+  extractFishName: ExtractFishName<E, Id>
+
+  /** Events to base the registration on */
+  events: Where<E>
```
Comment (aruediger) [inline src/types.ts:93]: should this be called something like `eventFilter` or just `filter` or `where`? (multiple places)

File: src/types.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+export type PondLike =
+  | {
+      observe: Pond['observe']
+    }
+  | Pond['observe']
+
+type FishToObservable = <S, E>(fish: Fish<S, E>) => Observable<S>
+
+/**
+ * Convert a plain Pond.observe function (or Pond) into one that returns RX6 Observables.
```
Comment (aruediger) [inline src/types.ts:16]:
```suggestion
 * Convert a plain Pond.observe function (or Pond) into one that returns RxJS v6 Observables.
```

File: src/types.ts [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+ * @internal
+ * @param pond pond or pond.observe
+ */
+export const extractObsFn = (pond: PondLike): Pond['observe'] =>
+  typeof pond === 'function' ? pond : pond.observe
+
+/**
+ * Observe a Fish as RXjs Obs
```
Comment (aruediger) [inline src/types.ts:24]:
```suggestion
 * Creates a converter from Fish to RxJS Observable.
```

File: src/types.ts [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+ * @internal
+ **/
+export const obs = (pond: PondLike): FishToObservable => {
+  const o = extractObsFn(pond)
+  return <S, E>(fish: Fish<S, E>) =>
+    new Observable<S>(subscriber => o(fish, s => subscriber.next(s)))
+}
+
+/** How to find a Fish’s "name" (to be remembered by the registry) from an event of type E. */
+export type ExtractFishName<E, Id> = (event: E) => Id
```
Comment (aruediger) [inline src/types.ts:34]:
```suggestion
export type ExtractFishName<E, Name> = (event: E) => Name
```

File: src/types.ts [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+/**
+ * Registry Fish State:
+ * Just an object that contains 'true' for every active Fish Name.
```
Comment (aruediger) [inline src/types.ts:38]:
```suggestion
 * An object that contains `true` for every active Fish Name.
```

File: README.md [hunk 12]
```diff
@@ -23,18 +23,35 @@ npm install @actyx-contrib/registry
...
-import { ChatFish, EventType } from "./fish/chatFish"
-import { Pond } from '@actyx/pond'
-
-Pond.default().then(pond => {
-  export const ChatRoomRegistryFish = createRegistryFish(ChatFish, EventType.message)
-  
-  observeRegistry(pond, ChatRoomRegistryFish, ChatFish)
-    .subscribe(console.log)
+import { FishType } from '@actyx-contrib/fish-type-v2'
+import { fromTypes } from '@actyx-contrib/registry-v2'
```
Comment (aruediger) [inline README.md:30]:
```suggestion
import { fromEventTypes } from '@actyx-contrib/registry-v2'
```

File: jest.config.js [hunk 13]
```diff
@@ -8,7 +8,7 @@ module.exports = {
   coverageThreshold: {
     global: {
       statements: 89,
-      branches: 84,
+      branches: 80,
```
Comment (aruediger) [inline jest.config.js:11]: generic registry wants test??

File: package.json [hunk 14]
```diff
@@ -24,29 +24,26 @@
   "author": "Actyx AG",
   "license": "Apache-2.0",
   "peerDependencies": {
-    "@actyx/pond": "~1.1",
-    "fp-ts": "~1.19.4",
-    "io-ts": "~1.10.1",
-    "io-ts-types": "~0.4.1",
-    "rxjs": "~5.5.12"
```
Comment (aruediger) [inline package.json:31]: 🙂

File: src/index.ts [hunk 15]
```diff
@@ -1,242 +1,18 @@
...
-  const registryFish = createRegistryFish(entityFish, addEventOrEventHandler, removeEvent)
-  return observeRegistry(pond, registryFish, entityFish)
+export const generic = {
+  create: Generic.createRegistryFish,
+  observeAll: Generic.observeAllGeneric,
 }
+
+export const observe = observeRegistry
+
+export { observeRegistryMap } from './observe'
```
Comment (aruediger) [inline src/index.ts:17]:
```suggestion
export { observeRegistryMap, observeRegistry as observe } from './observe'
```

File: src/index.ts [hunk 16]
```diff
@@ -1,242 +1,18 @@
...
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
-import { FishName, FishType, OnStateChange, Pond, Semantics, Subscription } from '@actyx/pond'
-import { Observable } from 'rxjs'
-import { combineLatest } from 'rxjs/observable/combineLatest'
+import * as FromEventTypes from './from-event-types'
+import * as Generic from './generic-registry'
+import { observeRegistry } from './observe'
```
Comment (aruediger) [inline src/index.ts:3]:
```suggestion
```

**Review by aruediger — COMMENTED** [review]
me likey!

File: jest.config.js (hunk 13 above)
Comment (benjamin-actyx) [reply jest.config.js:11]: Forgot to add the file :D

File: src/generic-registry.spec.ts [hunk 17]
```diff
@@ -0,0 +1,117 @@
+// import { Observable } from 'rxjs'
```
Comment (aruediger) [inline src/generic-registry.spec.ts:1]:
```suggestion
```

File: src/generic-registry.spec.ts [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+
+const emitEvent = async (pond: Pond, e: Event) =>
+    await pond
+        .emit<Event>(TestTag.withId(TestFish.extractFishName(e)), e)
+        .toPromise()
+        .then(() => new Promise(resolve => setTimeout(resolve, 0)))
+
+const nxt = async <T>(o: Observable<T>) => await take(1)(o).toPromise()
+
+describe('generic registry', () => {
```
Comment (aruediger) [inline src/generic-registry.spec.ts:50]: could you add tests for `createRegistryFish`, too? test coverage will be huuuge!

**Review by aruediger — COMMENTED** [review]
I pushed some lint fixes. Could you update the `example` folder, too (if it's still needed)?

File: src/generic-registry.spec.ts (hunk 18 above)
Comment (benjamin-actyx) [reply src/generic-registry.spec.ts:50]: Well it’s used by `observeAll`, isn’t it? I don’t understand why the coverage calculator complains :(

**Pull request closed without merging**